File: src/agent/agent.ts
/**
 * Autonomous Testing Agent
 * ========================
 *
 * Main entry point for running the autonomous testing agent loop.
 * Uses Claude Agent SDK to execute test planning and test execution sessions.
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { createSdkOptions } from "./client.ts";
import type { AgentOptions } from "./config.ts";
import { AUTO_CONTINUE_DELAY_MS, DEFAULT_MODEL } from "./config.ts";
import {
	getResumeResetPrompt,
	getTestExecutorPrompt,
	getTestPlannerPrompt,
	getTestReportPrompt,
	ProgressTracker,
	printTestProgressSummary,
	printTestSessionHeader,
	setupProjectDirectory,
	shouldResumeSession,
	TokenUsageTracker,
	updateSessionState,
} from "./services/index.ts";
import { PricingCalculator } from "./services/pricing.ts";
import { runAgentSession } from "./session.ts";
import { SessionStatus } from "./types/index.ts";
import { sleep } from "./utils/index.ts";

/**
 * Format duration from milliseconds to human-readable string (e.g., "6m 2s")
 */
export function formatDuration(totalMs: number): string {
	const totalSeconds = Math.floor(totalMs / 1000);
	const minutes = Math.floor(totalSeconds / 60);
	const seconds = totalSeconds % 60;

	if (minutes > 0) {
		return `${minutes}m ${seconds}s`;
	}
	return `${seconds}s`;
}

/**
 * Format token count to human-readable string (e.g., "1.34M", "500K")
 */
export function formatTokenCount(tokens: number): string {
	if (tokens >= 1_000_000) {
		return `${(tokens / 1_000_000).toFixed(2)}M`;
	}
	if (tokens >= 1_000) {
		return `${(tokens / 1_000).toFixed(0)}K`;
	}
	return tokens.toString();
}

/**
 * Escape HTML special characters to prevent XSS attacks.
 * Used when interpolating values into HTML templates.
 */
function escapeHtml(str: string): string {
	const htmlEscapes: Record<string, string> = {
		"&": "&amp;",
		"<": "&lt;",
		">": "&gt;",
		'"': "&quot;",
		"'": "&#39;",
	};
	return str.replace(/[&<>"']/g, (char) => htmlEscapes[char] || char);
}

// ============================================================================
// Console Output Helpers
// ============================================================================

const SEPARATOR_WIDTH = 70;

/**
 * Print a section header with double-line separators (=)
 */
function printSectionHeader(title: string): void {
	console.log(`\n${"=".repeat(SEPARATOR_WIDTH)}`);
	console.log(`  ${title}`);
	console.log("=".repeat(SEPARATOR_WIDTH));
}

/**
 * Print a subsection header with single-line separators (-)
 */
function printSubsectionHeader(title: string): void {
	console.log(`\n${"-".repeat(SEPARATOR_WIDTH)}`);
	console.log(`  ${title}`);
	console.log("-".repeat(SEPARATOR_WIDTH));
}

/**
 * Print a closing separator (=)
 */
function printSectionFooter(): void {
	console.log("=".repeat(SEPARATOR_WIDTH));
}

/**
 * Update HTML report with final cost statistics from usage_statistics.json.
 * This is called after the test_report session completes to ensure accurate costs.
 *
 * @param projectDir - Project directory path
 */
export async function updateHtmlReportCosts(projectDir: string): Promise<void> {
	try {
		// Read usage statistics
		const usageStatsFile = join(projectDir, "usage_statistics.json");
		if (!existsSync(usageStatsFile)) {
			console.log(
				"[Warning] usage_statistics.json not found, skipping HTML cost update",
			);
			return;
		}

		const usageStats = JSON.parse(readFileSync(usageStatsFile, "utf-8"));
		const summary = usageStats.summary;

		// Find the latest test-reports directory
		const testReportsDir = join(projectDir, "test-reports");
		if (!existsSync(testReportsDir)) {
			console.log(
				"[Warning] test-reports directory not found, skipping HTML cost update",
			);
			return;
		}

		// Look for Test_Report_Viewer.html directly in test-reports/ (flat structure per CLAUDE.md)
		const htmlReportFile = join(testReportsDir, "Test_Report_Viewer.html");

		if (!existsSync(htmlReportFile)) {
			console.log(
				"[Warning] Test_Report_Viewer.html not found, skipping cost update",
			);
			return;
		}

		// Read HTML content
		let htmlContent = readFileSync(htmlReportFile, "utf-8");

		// Calculate total duration from all sessions
		const totalDurationMs = usageStats.sessions.reduce(
			(sum: number, s: { durationMs: number }) => sum + s.durationMs,
			0,
		);

		// Prepare replacement values with HTML escaping to prevent XSS
		const totalCost = escapeHtml(`$${summary.totalCostUsd.toFixed(2)}`);
		const totalTokens = escapeHtml(formatTokenCount(summary.totalTokens));
		const totalDuration = escapeHtml(formatDuration(totalDurationMs));
		const totalSessions = escapeHtml(summary.totalSessions.toString());

		// Replace entire cost-grid section to handle any malformed AI-generated HTML
		// This is more robust than trying to match individual cost-value/cost-label pairs
		const costGridHtml = `<div class="cost-grid">
                    <div class="cost-item">
                        <div class="cost-value">${totalCost}</div>
                        <div class="cost-label">Total Cost</div>
                    </div>
                    <div class="cost-item">
                        <div class="cost-value">${totalTokens}</div>
                        <div class="cost-label">Total Tokens</div>
                    </div>
                    <div class="cost-item">
                        <div class="cost-value">${totalDuration}</div>
                        <div class="cost-label">Duration</div>
                    </div>
                    <div class="cost-item">
                        <div class="cost-value">${totalSessions}</div>
                        <div class="cost-label">Sessions</div>
                    </div>
                </div>`;

		// Match the entire cost-grid div and replace it
		// Structure: <div class="cost-grid">..items..</div></div></section>
		// (cost-grid closes, then cost-card closes, then section closes)
		htmlContent = htmlContent.replace(
			/<div class="cost-grid">[\s\S]*?<\/div>\s*<\/div>\s*<\/section>/,
			`${costGridHtml}\n            </div>\n        </section>`,
		);

		// Write updated HTML
		writeFileSync(htmlReportFile, htmlContent, "utf-8");

		console.log(`[Report Update] Updated HTML report costs:`);
		console.log(`  - Total Cost: ${totalCost}`);
		console.log(`  - Total Tokens: ${totalTokens}`);
		console.log(`  - Duration: ${totalDuration}`);
		console.log(`  - Sessions: ${totalSessions}`);
	} catch (error) {
		console.log(`[Warning] Failed to update HTML report costs: ${error}`);
	}
}

// ============================================================================
// Helper Functions for Agent Loop
// ============================================================================

/**
 * Print the agent banner with configuration info.
 */
function printAgentBanner(
	projectDir: string,
	model: string,
	maxIterations: number | null,
): void {
	printSectionHeader("AUTONOMOUS TESTING AGENT (TypeScript)");
	console.log(`\nProject directory: ${projectDir}`);
	console.log(`Model: ${model}`);
	if (maxIterations) {
		console.log(`Max iterations: ${maxIterations}`);
	} else {
		console.log("Max iterations: Unlimited (will run until completion)");
	}
	console.log();
}

// ============================================================================
// Main Agent Function
// ============================================================================

/**
 * Run the autonomous testing agent loop.
 *
 * @param options - Agent options
 * @returns Exit code: 0 for success, 1 for all tests blocked, 2 for idle loop detected
 */
export async function runAutonomousTestingAgent(
	options: AgentOptions,
): Promise<number> {
	const { projectDir, model = DEFAULT_MODEL, maxIterations = null } = options;

	// Print banner
	printAgentBanner(projectDir, model, maxIterations);

	// Create project directory
	if (!existsSync(projectDir)) {
		mkdirSync(projectDir, { recursive: true });
	}

	// Initialize token usage tracking with cache refresh
	const pricingCalculator = new PricingCalculator();
	// Pre-refresh pricing cache if stale (async fetch not possible in sync calculateCost)
	if (!pricingCalculator.isCacheValid()) {
		await pricingCalculator.updatePriceCache();
	}
	const usageTracker = new TokenUsageTracker(projectDir, pricingCalculator);

	// Check if this is a fresh start or continuation
	const testCasesFile = join(projectDir, "test_cases.json");
	let isFirstRun = !existsSync(testCasesFile);

	if (isFirstRun) {
		console.log("Fresh start - will use test planner agent");
		printSectionHeader("NOTE: First session takes 5-10 minutes!");
		console.log("  The agent is generating 50 detailed test cases.");
		console.log(
			"  This may appear to hang - it's working. Watch for [Tool: ...] output.",
		);
		printSectionFooter();
		console.log();
		setupProjectDirectory(projectDir);
	} else {
		console.log("Continuing existing testing project");
		await printTestProgressSummary(projectDir);
	}

	// Initialize progress tracker
	const progressTracker = new ProgressTracker(projectDir);

	// Main loop
	let iteration = 0;
	let allTestsBlocked = false;
	let idleLoopDetected = false;
	let consecutiveNoProgress = 0;
	const MAX_NO_PROGRESS_SESSIONS = 3;
	let previousNotRunCount: number | null = null;

	while (true) {
		iteration++;

		// Check max iterations
		if (maxIterations && iteration > maxIterations) {
			console.log(`\nReached max iterations (${maxIterations})`);
			console.log("To continue, run the script again without --max-iterations");
			break;
		}

		// Print session header
		printTestSessionHeader(iteration, isFirstRun);

		// Determine session type
		const currentSessionType = isFirstRun ? "test_planner" : "test_executor";

		// Check if we should resume the previous session (only for executor sessions)
		let resumeSessionId: string | undefined;
		let isResumedSession = false;

		if (!isFirstRun) {
			// Get current test status for resume decision
			const currentStats = await progressTracker.countTestCases();
			const { resumeSessionId: resumeId, reason } = shouldResumeSession(
				projectDir,
				currentStats.notRun,
			);

			if (resumeId) {
				resumeSessionId = resumeId;
				isResumedSession = true;
				console.log(`[Session Resume] ${reason}`);
			} else {
				console.log(`[Session] Fresh start: ${reason}`);
			}
		}

		// Create SDK options with conditional session resume
		const { options: sdkOptions } = await createSdkOptions({
			projectDir,
			model,
			...(resumeSessionId ? { resumeSessionId } : {}),
		});

		// Choose prompt based on session type
		let prompt: string;

		if (isFirstRun) {
			prompt = await getTestPlannerPrompt();
			isFirstRun = false;
		} else {
			prompt = await getTestExecutorPrompt();
			// If resuming, prepend reset instructions to avoid "mission accomplished" trap
			if (isResumedSession) {
				prompt = getResumeResetPrompt() + prompt;
			}
		}

		// Create abort controller
		const abortController = new AbortController();

		// Run session
		const { result } = await runAgentSession(
			sdkOptions,
			prompt,
			abortController,
		);

		const { status, usageData } = result;

		// Give processes time to terminate gracefully
		await sleep(500);

		// Record usage statistics if available
		if (usageData?.usage) {
			try {
				const sessionRecord = usageTracker.recordSession({
					sessionId: usageData.sessionId,
					sessionType: currentSessionType as "test_planner" | "test_executor",
					model,
					durationMs: usageData.durationMs,
					numTurns: usageData.numTurns,
					tokens: usageData.usage,
					sdkCostUsd: usageData.totalCostUsd,
				});
				usageTracker.displaySessionStats(sessionRecord);
			} catch (error) {
				console.log(`[Warning] Failed to record usage statistics: ${error}`);
			}
		} else {
			console.log("[Warning] No usage data available for this session");
		}

		// Update session state for conditional resume (executor sessions only)
		if (currentSessionType === "test_executor" && usageData?.sessionId) {
			const postSessionStats = await progressTracker.countTestCases();
			// Map session status to state update values
			// Note: updateSessionState only accepts "continue" | "context_overflow" | "error"
			// COMPLETED is mapped to "error" as a safe fallback (shouldn't happen in executor sessions)
			const sessionStatusMap: Record<
				SessionStatus,
				"continue" | "context_overflow" | "error"
			> = {
				[SessionStatus.CONTINUE]: "continue",
				[SessionStatus.CONTEXT_OVERFLOW]: "context_overflow",
				[SessionStatus.ERROR]: "error",
				[SessionStatus.COMPLETED]: "error",
			};
			const statusForState = sessionStatusMap[status];

			updateSessionState(
				projectDir,
				usageData.sessionId,
				statusForState,
				postSessionStats.notRun,
				isResumedSession,
			);
		}

		// Handle status
		if (status === SessionStatus.CONTINUE) {
			console.log(
				`\nAgent will auto-continue in ${AUTO_CONTINUE_DELAY_MS / 1000}s...`,
			);
			await printTestProgressSummary(projectDir);

			// Check if all tests are completed
			const stats = await progressTracker.countTestCases();

			if (stats.notRun === 0 && stats.total > 0) {
				if (stats.blocked === stats.total) {
					allTestsBlocked = true;
					printSectionHeader("ALL TESTS BLOCKED!");
					console.log(`\n  Total: ${stats.total}`);
					console.log(`  Blocked: ${stats.blocked}`);
					console.log("\n  Cannot proceed due to blocking issues.");
					printSectionFooter();
					break;
				}

				printSectionHeader("ALL TESTS COMPLETED!");
				console.log(`\n  Total: ${stats.total}`);
				console.log(`  Passed: ${stats.passed}`);
				console.log(`  Failed: ${stats.failed}`);
				console.log(`  Blocked: ${stats.blocked}`);
				console.log("\n  All test cases have been executed.");
				printSectionFooter();
				break;
			}

			// Detect idle sessions (no progress made)
			if (previousNotRunCount !== null) {
				if (stats.notRun === previousNotRunCount) {
					consecutiveNoProgress++;
					console.log(
						`\n[Warning] No progress made this session (${consecutiveNoProgress}/${MAX_NO_PROGRESS_SESSIONS} idle sessions)`,
					);

					if (consecutiveNoProgress >= MAX_NO_PROGRESS_SESSIONS) {
						idleLoopDetected = true;
						printSectionHeader("IDLE LOOP DETECTED!");
						console.log(
							`\n  ${MAX_NO_PROGRESS_SESSIONS} consecutive sessions with no test progress.`,
						);
						console.log(`  Remaining "Not Run" tests: ${stats.notRun}`);
						console.log("\n  Possible causes:");
						console.log(
							'    - Agent may be stuck due to "MISSION ACCOMPLISHED" in claude-progress.txt',
						);
						console.log("    - Blocking defects preventing test execution");
						console.log(
							"    - Environment issues preventing browser automation",
						);
						console.log("\n  Recommended actions:");
						console.log(
							"    1. Review claude-progress.txt and remove premature completion claims",
						);
						console.log(
							"    2. Check test_cases.json for remaining Not Run tests",
						);
						console.log("    3. Restart the agent after addressing issues");
						printSectionFooter();
						break;
					}
				} else {
					// Progress was made, reset counter
					consecutiveNoProgress = 0;
				}
			}
			previousNotRunCount = stats.notRun;

			await sleep(AUTO_CONTINUE_DELAY_MS);
		} else if (status === SessionStatus.CONTEXT_OVERFLOW) {
			console.log("\n[Context Overflow Recovery] Starting fresh session");
			await sleep(AUTO_CONTINUE_DELAY_MS);
		} else if (status === SessionStatus.ERROR) {
			console.log("\nSession encountered an error");
			console.log("Will retry with a fresh session...");
			await sleep(AUTO_CONTINUE_DELAY_MS);
		}

		// Small delay between sessions
		if (maxIterations === null || iteration < maxIterations) {
			console.log("\nPreparing next session...\n");
			await sleep(1000);
		}
	}

	// Final summary
	printSectionHeader("TESTING SESSION COMPLETE");
	console.log(`\nProject directory: ${projectDir}`);
	await printTestProgressSummary(projectDir);

	// Run report agent to generate final reports (only if tests were executed)
	const finalStats = await progressTracker.countTestCases();
	if (finalStats.total > 0 && finalStats.notRun === 0 && !idleLoopDetected) {
		printSectionHeader("GENERATING FINAL REPORTS");

		try {
			// Create SDK options for report session
			const { options: reportSdkOptions } = await createSdkOptions({
				projectDir,
				model,
			});

			// Get report prompt
			const reportPrompt = await getTestReportPrompt();

			// Create abort controller
			const reportAbortController = new AbortController();

			// Run report session
			const { result: reportResult } = await runAgentSession(
				reportSdkOptions,
				reportPrompt,
				reportAbortController,
			);

			// Record usage statistics for report session
			if (reportResult.usageData?.usage) {
				try {
					const reportSessionRecord = usageTracker.recordSession({
						sessionId: reportResult.usageData.sessionId,
						sessionType: "test_report",
						model,
						durationMs: reportResult.usageData.durationMs,
						numTurns: reportResult.usageData.numTurns,
						tokens: reportResult.usageData.usage,
						sdkCostUsd: reportResult.usageData.totalCostUsd,
					});
					usageTracker.displaySessionStats(reportSessionRecord);
				} catch (error) {
					console.log(
						`[Warning] Failed to record report session usage: ${error}`,
					);
				}
			}

			console.log("\n[Report Generation] Complete");

			// Update HTML report with final cost statistics (including test_report session)
			await updateHtmlReportCosts(projectDir);
		} catch (error) {
			console.log(`\n[Warning] Failed to generate reports: ${error}`);
		}
	}

	// Print instructions
	printSubsectionHeader("TO VIEW TEST REPORTS:");
	console.log(`\n  cd ${projectDir}/test-reports`);
	console.log("  # Open the HTML report viewer in a browser");
	console.log(`${"-".repeat(SEPARATOR_WIDTH)}`);

	console.log("\nDone!");

	// Exit codes: 0 = success, 1 = all tests blocked, 2 = idle loop detected
	if (idleLoopDetected) {
		return 2;
	}
	return allTestsBlocked ? 1 : 0;
}

File: src/agent/agent.py
"""
Agent Session Logic
===================

Core agent interaction functions for running autonomous coding sessions.
"""

import asyncio
import time
from enum import Enum
from pathlib import Path
from typing import Optional, Tuple

from claude_agent_sdk import ClaudeSDKClient

from client import create_client
from progress import (
    print_test_session_header,
    print_test_progress_summary,
    count_test_cases,
)
from prompts import (
    get_test_planner_prompt,
    get_test_executor_prompt,
    copy_templates_to_project,
    copy_utils_to_project,
    copy_test_spec_to_project,
)
from token_usage import TokenUsageTracker, PricingCalculator, CostReportGenerator, update_html_report_cost_statistics


# Configuration Constants
AUTO_CONTINUE_DELAY_SECONDS = 3


class SessionStatus(Enum):
    """Status of an agent session."""
    CONTINUE = "continue"
    ERROR = "error"


def format_tool_use_output(
    tool_name: str,
    thinking_time: float,
    tool_input: Optional[object] = None,
    max_input_len: int = 200
) -> None:
    """
    Format and print tool use information.

    Args:
        tool_name: Name of the tool being used
        thinking_time: Time elapsed since last event
        tool_input: Optional tool input to display
        max_input_len: Maximum length of input string to display
    """
    print(f"\n[Tool: {tool_name}] (after {thinking_time:.1f}s thinking)", flush=True)

    if tool_input is not None:
        input_str = str(tool_input)
        if len(input_str) > max_input_len:
            print(f"   Input: {input_str[:max_input_len]}...", flush=True)
        else:
            print(f"   Input: {input_str}", flush=True)


def format_tool_result_output(
    result_content: str,
    is_error: bool,
    execution_time: Optional[float] = None,
    max_error_len: int = 500
) -> None:
    """
    Format and print tool result information.

    Args:
        result_content: The content of the tool result
        is_error: Whether the result is an error
        execution_time: Optional execution time to display
        max_error_len: Maximum length of error string to display
    """
    time_suffix = f" (took {execution_time:.1f}s)" if execution_time is not None else ""

    # Check if command was blocked by security hook
    if "blocked" in str(result_content).lower():
        print(f"   [BLOCKED]{time_suffix} {result_content}", flush=True)
    elif is_error:
        # Show errors (truncated)
        error_str = str(result_content)[:max_error_len]
        print(f"   [Error]{time_suffix} {error_str}", flush=True)
    else:
        # Tool succeeded - just show brief confirmation
        print(f"   [Done]{time_suffix}", flush=True)


async def run_agent_session(
    client: ClaudeSDKClient,
    message: str,
    project_dir: Path,
) -> Tuple[SessionStatus, str, Optional[dict]]:
    """
    Run a single agent session using Claude Agent SDK.

    Args:
        client: Claude SDK client
        message: The prompt to send
        project_dir: Project directory path

    Returns:
        Tuple of (status, response_text, usage_data) where:
        - status indicates whether to continue or if an error occurred
        - response_text is the agent's text response
        - usage_data contains token usage information (or None if unavailable)
    """
    print("Sending prompt to Claude Agent SDK...\n")

    try:
        # Send the query and measure time
        start_time = time.time()
        await client.query(message)
        query_time = time.time() - start_time
        print(f"[Query sent in {query_time:.1f}s]\n")

        # Collect response text and show tool use
        response_text = ""
        last_event_time = time.time()
        tool_start_time: Optional[float] = None
        usage_data = None

        async for msg in client.receive_response():
            msg_type = type(msg).__name__

            # Handle ResultMessage (contains usage information)
            if msg_type == "ResultMessage":
                usage_data = {
                    "usage": getattr(msg, "usage", None),
                    "total_cost_usd": getattr(msg, "total_cost_usd", None),
                    "duration_ms": getattr(msg, "duration_ms", 0),
                    "num_turns": getattr(msg, "num_turns", 0),
                    "session_id": getattr(msg, "session_id", "unknown"),
                }

            # Handle AssistantMessage (text and tool use)
            elif msg_type == "AssistantMessage" and hasattr(msg, "content"):
                for block in msg.content:
                    block_type = type(block).__name__

                    if block_type == "TextBlock" and hasattr(block, "text"):
                        response_text += block.text
                        print(block.text, end="", flush=True)
                    elif block_type == "ToolUseBlock" and hasattr(block, "name"):
                        # Calculate thinking time and format output
                        thinking_time = time.time() - last_event_time
                        tool_input = getattr(block, "input", None)
                        format_tool_use_output(block.name, thinking_time, tool_input)
                        tool_start_time = time.time()

            # Handle UserMessage (tool results)
            elif msg_type == "UserMessage" and hasattr(msg, "content"):
                for block in msg.content:
                    if type(block).__name__ == "ToolResultBlock":
                        result_content = getattr(block, "content", "")
                        is_error = getattr(block, "is_error", False)

                        # Calculate tool execution time
                        execution_time = None
                        if tool_start_time is not None:
                            execution_time = time.time() - tool_start_time

                        format_tool_result_output(result_content, is_error, execution_time)

                        # Update last event time for next thinking time calculation
                        last_event_time = time.time()
                        tool_start_time = None

        print("\n" + "-" * 70 + "\n")
        return SessionStatus.CONTINUE, response_text, usage_data

    except Exception as e:
        print(f"Error during agent session: {e}")
        return SessionStatus.ERROR, str(e), None


async def run_autonomous_testing_agent(
    project_dir: Path,
    model: str,
    max_iterations: Optional[int] = None,
) -> int:
    """
    Run the autonomous testing agent loop.

    Args:
        project_dir: Directory for the testing project
        model: Claude model to use
        max_iterations: Maximum number of iterations (None for unlimited)

    Returns:
        Exit code: 0 for success, 1 for all tests blocked, 2 for other failures
    """
    print("\n" + "=" * 70)
    print("  AUTONOMOUS TESTING AGENT DEMO")
    print("=" * 70)
    print(f"\nProject directory: {project_dir}")
    print(f"Model: {model}")
    if max_iterations:
        print(f"Max iterations: {max_iterations}")
    else:
        print("Max iterations: Unlimited (will run until completion)")
    print()

    # Create project directory
    project_dir.mkdir(parents=True, exist_ok=True)

    # Initialize token usage tracking
    pricing_calculator = PricingCalculator()
    usage_tracker = TokenUsageTracker(project_dir, pricing_calculator)

    # Check if this is a fresh start or continuation
    test_cases_file = project_dir / "test_cases.json"
    is_first_run = not test_cases_file.exists()

    if is_first_run:
        print("Fresh start - will use test planner agent")
        print()
        print("=" * 70)
        print("  NOTE: First session takes 5-10 minutes!")
        print("  The agent is generating 50 detailed test cases.")
        print("  This may appear to hang - it's working. Watch for [Tool: ...] output.")
        print("=" * 70)
        print()
        # Copy the test spec into the project directory for the agent to read
        copy_test_spec_to_project(project_dir)
        copy_templates_to_project(project_dir)
        copy_utils_to_project(project_dir)
    else:
        print("Continuing existing testing project")
        print_test_progress_summary(project_dir)

    # Main loop
    iteration = 0
    all_tests_blocked = False  # Track if all tests are blocked

    while True:
        iteration += 1

        # Check max iterations
        if max_iterations and iteration > max_iterations:
            print(f"\nReached max iterations ({max_iterations})")
            print("To continue, run the script again without --max-iterations")
            break

        # Print session header
        print_test_session_header(iteration, is_first_run)

        # Create client (fresh context)
        client = create_client(project_dir, model)

        # Choose prompt based on session type
        current_session_type = "test_planner" if is_first_run else "test_executor"
        if is_first_run:
            prompt = get_test_planner_prompt()
            is_first_run = False  # Only use test planner once
        else:
            prompt = get_test_executor_prompt()

        # Run session with async context manager
        async with client:
            status, response, usage_data = await run_agent_session(client, prompt, project_dir)

        # Explicitly cleanup client to ensure MCP server processes are terminated
        try:
            await client.cleanup()
            print("[Cleanup] Client resources released")
        except Exception as e:
            print(f"[Cleanup] Warning: {e}")

        # Give processes time to terminate gracefully
        await asyncio.sleep(0.5)

        # Record usage statistics if available
        if usage_data and usage_data.get("usage"):
            try:
                session_record = usage_tracker.record_session(
                    session_id=usage_data["session_id"],
                    session_type=current_session_type,
                    model=model,
                    duration_ms=usage_data["duration_ms"],
                    num_turns=usage_data["num_turns"],
                    tokens=usage_data["usage"],
                )
                usage_tracker.display_session_stats(session_record)
            except Exception as e:
                print(f"[Warning] Failed to record usage statistics: {e}")
        else:
            print("[Warning] No usage data available for this session")

        # Handle status
        if status == SessionStatus.CONTINUE:
            print(f"\nAgent will auto-continue in {AUTO_CONTINUE_DELAY_SECONDS}s...")
            print_test_progress_summary(project_dir)

            # Check if all tests are completed
            stats = count_test_cases(project_dir)

            if stats['not_run'] == 0 and stats['total'] > 0:
                # Check if all tests are blocked (critical blocker scenario)
                if stats['blocked'] == stats['total']:
                    all_tests_blocked = True
                    print("\n" + "=" * 70)
                    print("  ⚠️  ALL TESTS BLOCKED!")
                    print("=" * 70)
                    print(f"\n  Total: {stats['total']}")
                    print(f"  Blocked: {stats['blocked']}")
                    print("\n  ❌ Cannot proceed due to blocking issues.")
                    print("  All test cases are blocked by infrastructure or dependencies.")
                    print("  Review defect reports and resolve blockers before retrying.")
                    print("=" * 70)
                    break

                # Normal completion: at least some tests passed or failed
                print("\n" + "=" * 70)
                print("  ALL TESTS COMPLETED!")
                print("=" * 70)
                print(f"\n  Total: {stats['total']}")
                print(f"  Passed: {stats['passed']}")
                print(f"  Failed: {stats['failed']}")
                print(f"  Blocked: {stats['blocked']}")
                print("\n  All test cases have been executed.")
                print("  Final reports should have been generated by the agent.")
                print("=" * 70)
                break

            await asyncio.sleep(AUTO_CONTINUE_DELAY_SECONDS)

        elif status == SessionStatus.ERROR:
            print("\nSession encountered an error")
            print("Will retry with a fresh session...")
            await asyncio.sleep(AUTO_CONTINUE_DELAY_SECONDS)

        # Small delay between sessions
        if max_iterations is None or iteration < max_iterations:
            print("\nPreparing next session...\n")
            await asyncio.sleep(1)

    # Final summary
    print("\n" + "=" * 70)
    print("  TESTING SESSION COMPLETE")
    print("=" * 70)
    print(f"\nProject directory: {project_dir}")
    print_test_progress_summary(project_dir)

    # Generate cost report
    try:
        report_generator = CostReportGenerator(usage_tracker)
        cost_report = report_generator.generate_markdown_report()

        # Find latest test-reports directory
        test_reports_dir = project_dir / "test-reports"
        if test_reports_dir.exists():
            report_dirs = sorted(test_reports_dir.glob("*"), key=lambda p: p.stat().st_mtime)
            if report_dirs:
                latest_report_dir = report_dirs[-1]
                cost_report_path = latest_report_dir / "cost_statistics.md"
                cost_report_path.write_text(cost_report)
                print(f"\n[Cost Report] Saved to: {cost_report_path}")
            else:
                # No test reports yet, save to project root
                cost_report_path = project_dir / "cost_statistics.md"
                cost_report_path.write_text(cost_report)
                print(f"\n[Cost Report] Saved to: {cost_report_path}")
        else:
            # No test-reports directory, save to project root
            cost_report_path = project_dir / "cost_statistics.md"
            cost_report_path.write_text(cost_report)
            print(f"\n[Cost Report] Saved to: {cost_report_path}")
    except Exception as e:
        print(f"\n[Warning] Failed to generate cost report: {e}")

    # Post-process HTML report to update cost statistics
    try:
        update_html_report_cost_statistics(project_dir)
    except Exception as e:
        print(f"\n[Warning] Failed to update HTML report cost statistics: {e}")

    # Print instructions for viewing test reports
    print("\n" + "-" * 70)
    print("  TO VIEW TEST REPORTS:")
    print("-" * 70)
    print(f"\n  cd {project_dir.resolve()}/test-reports")
    print("  # Open the HTML report viewer in a browser")
    print("  # Or browse the markdown reports in test-case-reports/ and defect-reports/")
    print("  # View cost statistics in cost_statistics.md")
    print("-" * 70)

    print("\nDone!")

    # Return appropriate exit code
    if all_tests_blocked:
        return 1  # All tests blocked
    return 0  # Success
